targets/arm/mikroe/sam/src/gpio/implementations/implementation_1/hal_ll_gpio_port.c

```diff
@@ -0,0 +1,299 @@
+**
+** Commercial License Usage
+**
+** Licensees holding valid commercial NECTO compilers AI licenses may use this
+** file in accordance with the commercial license agreement provided with the
+** Software or, alternatively, in accordance with the terms contained in
+** a written agreement between you and The MikroElektronika Company.
+** For licensing terms and conditions see
+** https://www.mikroe.com/legal/software-license-agreement.
+** For further information use the contact form at
+** https://www.mikroe.com/contact.
+**
+**
+** GNU Lesser General Public License Usage
+**
+** Alternatively, this file may be used for
+** non-commercial projects under the terms of the GNU Lesser
+** General Public License version 3 as published by the Free Software
+** Foundation: https://www.gnu.org/licenses/lgpl-3.0.html.
+**
+** The above copyright notice and this permission notice shall be
+** included in all copies or substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+** OF MERCHANTABILITY, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+** TO THE WARRANTIES FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+** IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+** DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT
+** OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
+** OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+**
+****************************************************************************/
+/*!
+ * @file  hal_ll_gpio_port.c
+ * @brief GPIO HAL LOW LEVEL ported layer implementation.
+ */
+
+#include "hal_ll_gpio_port.h"
+#include "hal_ll_pmc.h"
+
+// ------------------------------------------------------------- PRIVATE MACROS
+
+/**< @brief Utility macro for geting pin name based on port and bit for that pin on that port. */
+#define HAL_LL_GET_PIN(port_name, pin_index) ( ( hal_ll_pin_name_t ) ( ( port_name << 0x04 ) + pin_index ) )
+
+// -------------------------------------------------------------- PRIVATE TYPES
+
+/**
+ *  Defines used in source
+ */
+ // GPIO Register addresses and offsets
+#ifdef __PORT_A_CN
+#define HAL_LL_GPIOA_BASE_ADDR 0x400E0E00UL
+#define HAL_LL_PMC_ID_PIOA_BIT 0x0A
+#endif
+#ifdef __PORT_B_CN
+#define HAL_LL_GPIOB_BASE_ADDR 0x400E1000UL
+#define HAL_LL_PMC_ID_PIOB_BIT 0x0B
+#endif
+#ifdef __PORT_C_CN
+#define HAL_LL_GPIOC_BASE_ADDR 0x400E1200UL
+#define HAL_LL_PMC_ID_PIOC_BIT 0x0C
+#endif
+#ifdef __PORT_D_CN
+#define HAL_LL_GPIOD_BASE_ADDR 0x400E1400UL
+#define HAL_LL_PMC_ID_PIOD_BIT 0x10
+#endif
+#ifdef __PORT_E_CN
+#define HAL_LL_GPIOE_BASE_ADDR 0x400E1600UL
+#define HAL_LL_PMC_ID_PIOE_BIT 0x11
+#endif
+
+
+#define HAL_LL_PERIPHERAL_A 0
+#define HAL_LL_PERIPHERAL_B 1
+#define HAL_LL_PERIPHERAL_C 2
+#define HAL_LL_PERIPHERAL_D 3
+// EOF GPIO Register addresses and offsets
+
+
+/*!< @brief GPIO PORT array */
+static const uint32_t _hal_ll_gpio_port_base [ 5 ] =
```

Comment:
Maybe use `PORT_COUNT` instead of hard-coding a value.

Reply:
Agreed. Changed it.